Hexagon: sfmax/sfmin instructions are IEEE754-2019

wzssyqa opened the pull request:
The min/max instructions of Hexagon follow IEEE754-2019 minimumNumber/maximumNumber,
aka 
   FMINIMUMNUM and FMAXIMUMNUM
instead of
  FMAXNUM and FMINNUM

wzssyqa committed:
Hexagon: sfmax/sfmin instructions are IEEE754-2019

The min/max instructions of Hexagon follow IEEE754-2019, aka
  FMINIMUMNUM and FMAXIMUMNUM
instead of
  FMAXNUM and FMINNUM

androm3da commented:
Can you help me understand the distinction between these?

Also: is it safe to remove `fm{ax,in}num`?  Should this be an addition instead of a replacement?  Or does one of these imply/satisfy the other?

arsenm commented on llvm/lib/Target/Hexagon/HexagonISelLowering.cpp:
```diff
@@ -1639,28 +1640,75 @@ HexagonTargetLowering::HexagonTargetLowering(const TargetMachine &TM,
-    ISD::EXTRACT_SUBVECTOR,     ISD::INSERT_SUBVECTOR,
-    ISD::CONCAT_VECTORS,        ISD::VECTOR_SHUFFLE,
-    ISD::SPLAT_VECTOR,
+      // Integer arithmetic:
```
Avoid reformatting this whole table

aankit-ca commented:
> Can you help me understand the distinction between these?
> 
> Also: is it safe to remove `fm{ax,in}num`? Should this be an addition instead of a replacement? Or does one of these imply/satisfy the other?

I guess the distinction between FMAXIMUMNUM and FMAXIMUM is in the handling of snan's. FMAXIMUMNUM variant does not distinguish between snan's and qnan's which is what the sfmax instruction does as well. Is my understanding correct @wzssyqa ?

arsenm commented:
> Can you help me understand the distinction between these?

Mostly signaling nan handling. The 2008 operation nonsensically behaves in a way opposite of how it does with a qnan input. We have this chart now https://llvm.org/docs/LangRef.html#llvm-min-intrinsics-comparation

> 
> Also: is it safe to remove `fm{ax,in}num`? Should this be an addition instead of a replacement? Or does one of these imply/satisfy the other?

All 3 pairs of intrinsics should be independently supported, and you can use one set to implement the others.

The ISA documentation for these operands is universally poor. The document I found for sfmin states:
"Minimum of two floating point values. If one value is a NaN, the other is chosen."

So if taken at face value, this is correct. The title and description should clarify which operation, as 2019 has both minimum and minimumNumber. This does explicitly not state if -0 is ordered as less than 0, but I'm assuming it is as that's the 2019 behavior for both operations, and also I haven't seen an example of an ISA that didn't have the correct behavior

arsenm commented on llvm/test/CodeGen/Hexagon/fminmax-v67.ll:
```diff
@@ -2,7 +2,7 @@
 
 ; CHECK-LABEL: t1
-; CHECK: dfmax
+; CHECK-NOT: dfmax
```
-NOT checks are next to useless, this should have positive checks

arsenm commented on llvm/test/CodeGen/Hexagon/fminmax-v67.ll:
```diff
@@ -2,7 +2,7 @@
 
 
 ; CHECK-LABEL: t1
```
Missing test coverage for other targets?

androm3da replied:
That problem was on the baseline.  But I agree: while we're here, it's a great idea to add a few more `RUN` lines.

wzssyqa replied:
I have feel knowledge about the flavors of Hexagon.
So which to add?

arsenm replied:
Add new run lines in a separate change

wzssyqa replied:
Is this patch OK?

wzssyqa committed:
Fix code style

Avoid reformatting this whole table
-NOT checks are next to useless, this should have positive checks

wzssyqa committed:
Merge branch 'main' into Hexgon_FMINNUM

wzssyqa commented:
ping

aankit-ca reviewed (commented):
I'm fine with the changes. @androm3da do you have any further comments for this PR?

androm3da commented:
> I'm fine with the changes. @androm3da do you have any further comments for this PR?

No comments, LGTM
